### PR TITLE
network: TC- Introduce Quick Fair Queueing (QFQ) 

### PR DESCRIPTION
**Pull request by ssahani**

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -3042,6 +3042,25 @@
     </variablelist>
   </refsect1>
 
+  <refsect1>
+    <title>[QuickFairScheduler] Section Options</title>
+    <para>The <literal>[QuickFairScheduler]</literal> section manages the queueing discipline
+    (qdisc) of Quick Fair Scheduler (QFQ).</para>
```
Comment (yuwata) [inline man/systemd.network.xml]: QFQ is "Quick Fair Queueing (Plus) Scheduler".

File: man/systemd.network.xml [hunk 2]
```diff
@@ -3042,6 +3042,25 @@
...
+
+    <variablelist class='network-directives'>
+      <xi:include href="tc.xml" xpointer="qdisc-parent" />
+      <xi:include href="tc.xml" xpointer="qdisc-handle" />
+
+      <varlistentry>
+        <term><varname>Weight=</varname></term>
```
Comment (yuwata) [inline man/systemd.network.xml:3276]: Weight is a property for the qfq class, not for qdisc.

File: man/systemd.network.xml (hunk 1 above)
Comment (ssahani) [reply man/systemd.network.xml]: Plus is something different. https://access.redhat.com/solutions/1459623

File: man/systemd.network.xml (hunk 2 above)
Comment (ssahani) [reply man/systemd.network.xml:3276]: ohh yes . I totally misread it . removed. thanks for the review.

**Comment by poettering** [comment]
looks superficially ok to me, but CI doesn't like it, and i'll leave this one to @yuwata to merge I guess

**Comment by ssahani** [comment]
CI passed now. ptal

File: man/systemd.network.xml (hunk 1 above)
Comment (yuwata) [reply man/systemd.network.xml]: OK. Regarding https://en.wikipedia.org/wiki/Network_scheduler#Algorithms, let's rename this to "Quick Fair Queueing", as "Scheduler" is redundant and applicable for all other algorithms.

**Comment by yuwata** [comment]
@ssahani I will rename "Quick Fair Scheduler" -> "Quick Fair Queueing", and add tests for this. Can I modify this PR as usual?

**Comment by ssahani** [comment]
> @ssahani I will rename "Quick Fair Scheduler" -> "Quick Fair Queueing", and add tests for this. Can I modify this PR as usual?

yes please thanks. No need to ask  :)

**Comment by yuwata** [comment]
Updated.
- also introduces [QuickFairQueueingClass] section,
- tests for QFQ scheduler are added.

PTAL.

File: src/network/tc/qfq.c [hunk 3]
```diff
@@ -0,0 +1,10 @@
+/* SPDX-License-Identifier: LGPL-2.1+
+ * Copyright © 2020 VMware, Inc. */
```
Comment (keszybz) [inline src/network/tc/qfq.c:2]: Is this absolutely required?

File: man/systemd.network.xml [hunk 4]
```diff
@@ -3263,6 +3263,34 @@
...
+    Quick Fair Queueing (qfq).</para>
+
+    <variablelist class='network-directives'>
+      <xi:include href="tc.xml" xpointer="tclass-parent" />
+      <xi:include href="tc.xml" xpointer="tclass-classid" />
+
+      <varlistentry>
+        <term><varname>Weight=</varname></term>
+        <listitem>
+          <para>Specifies the weight of the class. Takse an integer ranges 1 to 1023. Defaults to unset and kernel's
```
Comment (keszybz) [inline man/systemd.network.xml]: "Takes an integer in the range 1..1023."
"Defaults to unset, in which case the kernel default is used."

File: man/systemd.network.xml [hunk 5]
```diff
@@ -3263,6 +3263,34 @@
...
+          default is used.</para>
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>MaxPacketSize=</varname></term>
+        <listitem>
+          <para>Specifies the maximum packet size in bytes for the class. When suffixed with K, M, or G, the specified
+          size is parsed as Kilobytes, Megabytes, or Gigabytes, respectively, to the base of 1000. When unset,
+          kernel's default is used.</para>
```
Comment (keszybz) [inline man/systemd.network.xml]: "the kernel default"

File: src/network/tc/qfq.c (hunk 3 above)
Comment (yuwata) [reply src/network/tc/qfq.c:2]: cc @ssahani.

Comment (ssahani) [reply src/network/tc/qfq.c:2]: yes

**Comment by yuwata** [comment]
@keszybz Thank you for the review. I've updated the man page. PTAL.

**Comment by keszybz** [comment]
LGTM.

**Comment by keszybz** [comment]
semaphoreci failed with "Failed to download http://images.linuxcontainers.org//images/debian/buster/amd64/default/20200603_05:24//rootfs.tar.xz", looks unrelated.

File: man/systemd.network.xml [hunk 6]
```diff
@@ -3252,6 +3252,45 @@
...
+    Quick Fair Queueing (qfq).</para>
+
+    <variablelist class='network-directives'>
+      <xi:include href="tc.xml" xpointer="tclass-parent" />
+      <xi:include href="tc.xml" xpointer="tclass-classid" />
+
+      <varlistentry>
+        <term><varname>Weight=</varname></term>
+        <listitem>
+          <para>Specifies the weight of the class. Takse an integer in the range 1..1023. Defaults to
```
Comment (benjarobin) [inline man/systemd.network.xml:3278]: `Takse`: There is a typo here, if anybody wants to fix it

**Pull request merged**